File: leetcode/46. Permutations/permutations.js
function allPermutation(nums){
    let start = 0;
    let end = nums.length;
    let permutation = [];
        function backtrack(nums, start) {
            if (start == end) {
                permutation.push(nums.slice())
            }
            else {
            for (let i = start; i < end; i++) {
                [nums[i], nums[start]] = [nums[start], nums[i]];
                backtrack(nums, start+1);
                [nums[i], nums[start]] = [nums[start], nums[i]]
            }
        }
    }
    backtrack(nums, start)
    return permutation;
}

File: LeetCode/46. Permutations/permutations.cs
public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        int length = nums.Length - 1;
        var permutations = new List<IList<int>>();
        return DoPermute(nums, 0, length, permutations);
    }
    private static IList<IList<int>> DoPermute(int[] nums, int start, int end, List<IList<int>> list) {
        if (start == end)
            list.Add(new List<int>(nums));
        else {
            for (int i = start; i <= end; i++) {
                Swap(ref nums[i], ref nums[start]);
                DoPermute(nums, start + 1, end, list);
                Swap(ref nums[i], ref nums[start]);
            }
        }
        return list;
    }
    private static void Swap(ref int a, ref int b) {
        int temp = a;
        a = b;
        b = temp;
    }
}
